python/tchannel/io.py

```diff
@@ -19,8 +19,169 @@
 # THE SOFTWARE.
 
 from __future__ import absolute_import
+import os
+from .exceptions import StreamingException
+from .messages.common import StreamState
+
 
 try:
     from cStringIO import StringIO as BytesIO
 except ImportError:  # pragma: no cover
     from io import BytesIO  # noqa
+
+import tornado
+import tornado.gen
+import tornado.ioloop
+from tornado.iostream import PipeIOStream, StreamClosedError
+
+try:
+    from tornado.locks import Condition
+except ImportError:  # pragma: no cover
+    from toro import Condition
+
+from .messages import common
+
+
+class BaseArgStream(object):
+
+    def __init__(self):
+        """Base Arg Stream Class
+
+        _state is used to track the state of the arg stream.
+
+        When object is created, it is in the "init" state.
+
+        When write func is called, it will transformed to "streaming" state.
+
+        When close func is called or reach the end of the stream,
+        it will jump to "completed" state no matter what state it is before.
+
+        If the stream is closed, it won't allow to write anymore.
+        If the stream is closed, it will return whatever still remains in the
+        stream buffer, after that it will just return empty.
+        """
+        self._state = StreamState.init
+
+    def read(self):
+        """Async read from internal stream buffer
+
+        if it reaches the end of the stream, it will return empty(""), caller
+        will depend on the return value to tell if it reaches the end of the
+        stream.
+
+        It doesn't support seek functionality, which means it will only read
+        the stream data once in single direction.
+
+        NOTE: The data which has been read will be discarded in the stream.
+
+        :return:chunk of bytes read from stream
+        """
+        raise NotImplementedError()
+
+    def write(self, chunk):
+        """Async write to internal stream buffer
+
+        :return if stream has been closed, it will raise StreamingException
+        """
+        raise NotImplementedError()
+
+    def close(self):
+        """Close the stream
+
+        It updates the _state to completed.
```

Comment:
Again, state changes should not be in the docstring. Also, why does this class declare the state if it doesn't read or write to it at all. Maybe leave it up to the implementations to decide if they want to track state using StreamState? Some may decide they have more complicated needs and use something else to track state.